middleware/cancel/cancel.go

```diff
@@ -0,0 +1,228 @@
+	defer end()
+	return c.inner.Invoke(ctx, req)
+}
+
+func (c *cancelProvider) Check(ctx p.Context, req p.CheckRequest) (p.CheckResponse, error) {
+	ctx, end := c.cancel(ctx, noTimeout)
+	defer end()
+	return c.inner.Check(ctx, req)
+}
+
+func (c *cancelProvider) Diff(ctx p.Context, req p.DiffRequest) (p.DiffResponse, error) {
+	ctx, end := c.cancel(ctx, noTimeout)
+	defer end()
+	return c.inner.Diff(ctx, req)
+}
+
+func (c *cancelProvider) Create(ctx p.Context, req p.CreateRequest) (p.CreateResponse, error) {
+	ctx, end := c.cancel(ctx, req.Timeout)
+	defer end()
+	return c.inner.Create(ctx, req)
+}
+
+func (c *cancelProvider) Read(ctx p.Context, req p.ReadRequest) (p.ReadResponse, error) {
+	ctx, end := c.cancel(ctx, noTimeout)
+	defer end()
+	return c.inner.Read(ctx, req)
+}
+
+func (c *cancelProvider) Update(ctx p.Context, req p.UpdateRequest) (p.UpdateResponse, error) {
+	ctx, end := c.cancel(ctx, req.Timeout)
+	defer end()
+	return c.inner.Update(ctx, req)
+}
+
+func (c *cancelProvider) Delete(ctx p.Context, req p.DeleteRequest) error {
+	ctx, end := c.cancel(ctx, req.Timeout)
+	defer end()
+	return c.inner.Delete(ctx, req)
+}
+
+func (c *cancelProvider) Construct(pctx p.Context, typ string, name string,
+	ctx *pulumi.Context, inputs pprovider.ConstructInputs, opts pulumi.ResourceOption) (pulumi.ComponentResource, error) {
+	pctx, end := c.cancel(pctx, noTimeout)
+	defer end()
+	return c.inner.Construct(pctx, typ, name, ctx, inputs, opts)
+}
+
+// A data structure which provides amortized O(1) insertion, removal, and draining.
+type inOutCache[T any] struct {
+	values     []*entry[T] // An unorderd list of stored values
+	tombstones []int       // An unordered list of empty slots in values
+	m          sync.Mutex
+	inDrain    bool
+}
+
+type entry[T any] struct {
+	evict func() bool
+	value T
+}
+
+// Insert a new element into the inOutCahce. The new element can be ejected by calling
+// `evict`. If the element was already drained or if `evict` was already called, then
+// `evict` will return true. Otherwise it returns false.
+func (h *inOutCache[T]) insert(t T) (evict func() (missing bool)) {
+	h.m.Lock()
+	defer h.m.Unlock()
+	var i int
+	if len(h.tombstones) == 0 {
+		i = len(h.values)
+	} else {
+		i = h.tombstones[len(h.tombstones)-1]
+		h.tombstones = h.tombstones[:len(h.tombstones)-1]
+	}
+
+	el := &entry[T]{
+		value: t,
+	}
+	el.evict = func() bool {
+		if !h.inDrain {
+			h.m.Lock()
+			defer h.m.Unlock()
+		}
+		gone := el.evict == nil
+		if gone {
+			return true
+		}
+		el.evict = nil
+		h.values[i] = nil
+		h.tombstones = append(h.tombstones, i)
+		return gone
+
+	}
+
+	if len(h.tombstones) == 0 {
+		h.values = append(h.values, el)
+	} else {
+		h.values[i] = el
+	}
+	return el.evict
+}
+
+// Remove all values from the inOutCache.
+func (h *inOutCache[T]) drain() []T {
+	h.m.Lock()
+	defer h.m.Unlock()
+	h.inDrain = true
+	defer func() { h.inDrain = false }()
+	values := []T{}
+	// We reset tombstones.
```

Comment:
Could the old cache `h` values just be returned along with a fresh cache, i.e. `return h.values, &inOutCache[T]`?

Reply:
No, because `h.values` is of type `[]*entry[T]`. Some of `h.values` will be `nil`. We also need to call `evict` on all values we remove to indicate that they have already been handled.

I don't see the benefit of returning a new cache. Since a cache is inherently mutable, there is no reason to work towards immutability. We have already allocated the space for `h.values`, so we might as well use it. 

Can you explain why this is desirable?

Reply:
That makes sense, I was just wondering if we wanted to have an upper bound on the size of the cache, or possibly reset it each time it was drained.